src/contracts/core/AVSDirectory.sol

```diff
@@ -52,73 +60,275 @@ contract AVSDirectory is
-            operator, 
-            operatorRegistrationDigestHash, 
+            operator,
+            calculateOperatorSetRegistrationDigestHash({
+                avs: msg.sender,
+                operatorSetIds: operatorSetIds,
+                salt: operatorSignature.salt,
+                expiry: operatorSignature.expiry
+            }),
             operatorSignature.signature
         );
 
-        // Set the operator as registered
-        avsOperatorStatus[msg.sender][operator] = OperatorAVSRegistrationStatus.REGISTERED;
+        // Mutate `operatorSaltIsSpent` to `true` to prevent future respending.
+        operatorSaltIsSpent[operator][operatorSignature.salt] = true;
+
+        _registerToOperatorSets(operator, msg.sender, operatorSetIds);
+    }
+
+    /**
+     * @notice Called by an operator to deregister from an operator set
+     *
+     * @param operator The operator to deregister from the operatorSets.
+     * @param avs The address of the AVS to deregister the operator from.
+     * @param operatorSetIds The IDs of the operator sets.
+     * @param operatorSignature the signature of the operator on their intent to deregister or empty if the operator itself is calling
+     *
+     * @dev if the operatorSignature is empty, the caller must be the operator
+     * @dev this will likely only be called in case the AVS contracts are in a state that prevents operators from deregistering
+     */
+    function forceDeregisterFromOperatorSets(
+        address operator,
+        address avs,
+        uint32[] calldata operatorSetIds,
+        ISignatureUtils.SignatureWithSaltAndExpiry memory operatorSignature
+    ) external override onlyWhenNotPaused(PAUSER_OPERATOR_REGISTER_DEREGISTER_TO_OPERATOR_SETS) {
+        if (operatorSignature.signature.length == 0) {
+            require(msg.sender == operator, InvalidOperator());
+        } else {
+            // Assert operator's signature has not expired.
+            require(operatorSignature.expiry >= block.timestamp, SignatureExpired());
+            // Assert operator's signature `salt` has not already been spent.
+            require(!operatorSaltIsSpent[operator][operatorSignature.salt], SaltSpent());
+
+            // Assert that `operatorSignature.signature` is a valid signature for operator set deregistrations.
+            EIP1271SignatureUtils.checkSignature_EIP1271(
+                operator,
+                calculateOperatorSetForceDeregistrationTypehash({
+                    avs: avs,
+                    operatorSetIds: operatorSetIds,
+                    salt: operatorSignature.salt,
+                    expiry: operatorSignature.expiry
+                }),
+                operatorSignature.signature
+            );
+
+            // Mutate `operatorSaltIsSpent` to `true` to prevent future respending.
+            operatorSaltIsSpent[operator][operatorSignature.salt] = true;
+        }
+        _deregisterFromOperatorSets(avs, operator, operatorSetIds);
```

Comment:
given that this function (`forceDeregisterFromOperatorSets`) has the same end result as `deregisterOperatorFromOperatorSets` should these maybe be combined? or this would just be too confusing with too many ways to call the same function?

Reply:
@ypatil12 Perhaps you could shed some light here, I remember us needing to introduce force deregistrations for some edge case.

Reply:
I think the functionality makes sense, just asking if the entrypoints should be combined or not 😄